Applications/Spire/Source/OrderImbalanceIndicator/FilteredOrderImbalanceIndicatorModel.cpp

```diff
@@ -0,0 +1,114 @@
+#include "Spire/OrderImbalanceIndicator/FilteredOrderImbalanceIndicatorModel.hpp"
+
+using Filter = Spire::FilteredOrderImbalanceIndicatorModel::Filter;
+using namespace Nexus;
+using namespace Spire;
+
+FilteredOrderImbalanceIndicatorModel::FilteredOrderImbalanceIndicatorModel(
+  std::shared_ptr<OrderImbalanceIndicatorModel> source_model,
+  std::vector<Filter> filters)
+  : m_source_model(std::move(source_model)),
+    m_filters(std::move(filters)) {}
+
+QtPromise<std::vector<Nexus::OrderImbalance>>
+    FilteredOrderImbalanceIndicatorModel::load(
+    const TimeInterval& interval) {
+  return m_source_model->load(interval).then([=] (const auto& imbalances) {
```

Comment:
```FilteredOrderImbalanceIndicatorModel``` needs to have ownership of the promise that the source model returns, otherwise you risk a dangling reference.

Reply:
Does the caller not take ownership?

Reply:
Add a test case as follows:

* Construct a test model T (the reified model used for testing) (as a shared_ptr).
* Construct a filtered model F (as a unique_ptr) and pass as an argument T.
* Call F->load(P) for some arbitrary value P.
* Call F.reset(), destroying the filtered model.
* Call T.pop_load() and assign it to O.
* Call O.set_result(R) for some arbitrary value R.
* Expected result, no crash.

Reply:
Do you mean to reset T (the test model), or F?

I have:

```
    auto T = std::make_shared<TestOrderImbalanceIndicatorModel>();
    auto F = std::make_unique<FilteredOrderImbalanceIndicatorModel>(T,
      std::vector<FilteredOrderImbalanceIndicatorModel::Filter>());
    F->load(TimeInterval::closed(from_time_t(0), from_time_t(1000)));
    F.reset();
    auto O = wait(T->pop_load());
    O->set_result({});
```

And it's working as expected, in that even while the filter model, F, is destroyed, the promise still completes.

Reply:
Okay, I see the problem.

Reply:
Also use more descriptive names than the ones I give for your actual unit test.

Reply:
Is it possible to add a method to see if a QtPromise has called its continuation function?

If I understand it, I need to keep a handle to the QtPromise returned by the source model, but if load(...) is called 10 times and takes a year to resolve each call, I'll need to store those 10 promises and clean them up when they've resolved. I could store each promise as a struct with the promise and an is_completed bool, but I know you prefer a 'single source of truth'.

Reply:
No, you will have a map of ```QtPromise``` objects, you add them into that map and then in the callback you remove it from the map.

    auto id = ++m_next_promise_id;
    auto some_promise = some_load(
      [=] {
        m_pending_promises.erase(id);
      });
    m_pending_promises.insert(std::make_pair(id, std::move(some_promise)));